File: LeetCode/1818. Minimum Absolute Sum Difference/solution.py
class Solution:
    def minAbsoluteSumDiff(self, nums1: List[int], nums2: List[int]) -> int:
        arr = sorted(nums1)
        total = sum(abs(x1-x2) for x1, x2 in zip(nums1,nums2))
        if total == 0:
            return 0
        ans = total
        n = len(nums1)
        for x1, x2 in zip(nums1,nums2):
            idx = bisect.bisect_left(arr,x2)
            if idx < n:
                ans = min(ans, total-abs(x1-x2)+abs(arr[idx]-x2))
            if idx-1 >= 0:
                ans = min(ans, total-abs(x1-x2)+abs(arr[idx-1]-x2))
        return ans%(10**9+7)

File: LeetCode/1818. Minimum Absolute Sum Difference/solution.cpp
class Solution {
public:
int minAbsoluteSumDiff(vector<int>& n1, vector<int>& n2) {
    long res = 0, gain = 0;
    set<int> s(begin(n1), end(n1));
    for (int i = 0; i < n1.size(); ++i) {
        long original = abs(n1[i] - n2[i]);
        res += original;
        // if (gain < original) 
        {
            auto it = s.lower_bound(n2[i]);
            if (it != end(s))
                gain = max(gain, original - abs(*it - n2[i]));
            if (it != begin(s))
                gain = max(gain, original - abs(*prev(it) - n2[i]));
        }
    }
    return (res - gain) % 1000000007;
}
};

class Solution {
public:
    int minAbsoluteSumDiff(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size(), MOD = 1e9+7;
        long long tot = 0, res = LONG_MAX;
        set<int> st(nums1.begin(), nums1.end());
        for (int i = 0; i < n; i++){
            tot += abs(nums1[i]-nums2[i]);
        }
        for (int i = 0; i < n; i++){
            auto it = st.lower_bound(nums2[i]);
            int delta = INT_MAX;
            if (it != st.end())
                delta = min(delta, *it - nums2[i]);
            if (it != st.begin())
                delta = min(delta, nums2[i] - *(--it));
            res = min(res, tot+delta-abs(nums1[i]-nums2[i]));

        }
        return res%MOD;
    }
};